### ISSUE TITLE
feat: Simplify expressions definition to enable better auto-complete in IPython + nvim

### ISSUE THREAD

**Issue by MarcoGorelli** [pola-rs/polars#16933]
### Description

I'd brought up some time ago that autocomplete / tab-complete is unreliable in IPython. I thought this might have been due to limitations in recognising namespaces (like `.dt`). However, comparing with the much simpler Narwhals, that seems to not the case:

Let's see what happens if I type `pl.col('a').dt.to` - what's suggested?

demo from IPython:

Narwhals (correct suggestion):

![Screenshot 2024-06-13 145708](https://github.com/pola-rs/polars/assets/33491632/0aea1614-2297-424d-9ac1-155418bef2c8)

Polars (incorrect suggestion):

![image](https://github.com/pola-rs/polars/assets/33491632/e070cee1-f7ea-4117-aec5-c5b0d9173506)

---

From nvim (with pylsp):

Narwhals (correct suggestion):

![image](https://github.com/pola-rs/polars/assets/33491632/0fd910bf-3317-49e7-a561-8be592753c7e)

Polars (no suggestion):

![image](https://github.com/pola-rs/polars/assets/33491632/a0a9f878-72af-434b-a0cb-3918408fab59)

---

I _suspect_ (though haven't yet been able to verify) that some of the magic dispatch mechanisms in Polars may be to blame? They work fine at runtime, but mess up editors

If that's the case, then I'd advocate for drastically simplifying things on the Python Polars side - if that means copy-and-pasting a few definitions, then OK, so be it. But in terms of user adoption, a better user experience with auto-complete would really pay dividends

---

Note that the namespaces aren't the only cases where the discrepancy appears. I won't post screenshots, but you can verify that typing `pl.col('a').fi` / `nw.col('a').fi` gives:
- IPython:
  - Narwhals: suggests `fill_null` and `filter`
  - Polars: no suggestion **EDIT**: this is now fixed in Polars (but the `pl.col('a').dt.to` issue is still present!)
- nvim:
  - this works for both 👍 

---

A solution _could_ be to say that IPython and pylsp are wrong...but given how many users they have (especially the former, almost the whole data science world uses some variation of it...) I think it'd be a pity

**Comment by MarcoGorelli** [comment]
Yeah pylance seems more advanced. As much as I'd like to say "everyone should just use vscode or pycharm", people don't necessarily have control over what they use (especially at work)

**Comment by MarcoGorelli** [comment]
from discussion: it's desirable to make this work, it's just not clear what exactly is the culprit

**Comment by krassowski** [comment]
I narrowed it down to jedi inference being mislead by the `sphinx_accessor` being assigned to `property`. Commenting these two lines out:

https://github.com/pola-rs/polars/blob/ecfb7b5e09fd074ede5bd270453237d982495328/py-polars/polars/expr/expr.py#L101-L102

makes it work 100%:

![Image](https://github.com/user-attachments/assets/873dcad9-2c20-4d7f-8ba8-7a16339a1abf)

Jedi/parso do not know that `BUILDING_SPHINX_DOCS` is `False` when performing static inference. There might be a way to tell it to not enter branches with `BUILDING_SPHINX_DOCS`, or to accept subclasses or `property`, which is what `sphinx_accessor` is:

https://github.com/pola-rs/polars/blob/ecfb7b5e09fd074ede5bd270453237d982495328/py-polars/polars/_utils/various.py#L414

**Comment by MarcoGorelli** [comment]
thanks a tonne @krassowski for looking into this! cc @alexander-beedie in case you have ideas for how to resolve this

**Comment by alexander-beedie** [comment]
@krassowski, thanks very much for narrowing this down so effectively 👍 
@MarcoGorelli... I think I actually have a solution!! 😎
